proposals/0448-type-variable-scoping.rst

```diff
@@ -414,7 +414,19 @@ Proposed Change Specification
    (Right now, pattern synonyms require all such quantifications to occur before any term arguments,
    but accepted proposal `#402`_ allows these quantifications to occur in any order in data constructors.)
 
-#. Any type variables mentioned in a type application are considered binding sites, shadowing any in-scope type variables.
+#. In accordance with the `Visibility Orthogonality Principle`_,
+   the rules that determine whether a variable occurrence is a binding site or a use site are not affected by the presence of a ``@``.
+   At the time of writing, the rules are as follows:
+
+   * Outside pattern signatures,
```

Comment:
I think we must be very careful around the term "pattern signature", because different people have different understandings of this term, and this has led to confusion elsewhere in this space.

I claim that a pattern signature is the `ty` in the grammar form `(pat :: ty)`, when that occurs in a pattern (as it must).

However, I have seen it claimed that the ty in `data T (a :: ty)` is a pattern signature. The example in this PR describes the `ki` in `(ty :: ki)` as a pattern signature, when it occurs in a visible type application in a pattern. These cases are not, to me, signatures on patterns, and are thus not pattern signatures.

One relevant nuance here is that these two are grammatically different:

```
x1 (MkT (a :: ty))
x2 (MkT @(a :: ty))
```

In `x1`, the `a` is a pattern, and thus `ty` is a pattern signature. In `x2`, the `a` is a type -- not a pattern, as we don't have type patterns in our grammar -- and thus (to me) the `ty` in `x2` is not a pattern signature. Perhaps the VOP suggests we should introduce type patterns in our grammar in order to avoid this nuance, but I don't think that doing so is worthwhile.

I'm not disagreeing with what I understand as the intended substance of this proposal, but I do object to the use of "pattern signature" to describe something that is not grammatically a signature of a pattern.

Reply:
Would it be acceptable to say "pattern or kind signatures"?

Reply:
Yes, I think so.

Reply:
Actually, those two paragraphs refer to pattern signatures specifically, since they describe the existing rules (the preceding sentence is "at the moment of writing, the rules are..."). Therefore, no change required. It is implied that we take the existing rules and generalize them in such a way that VOP is made to hold. For clarity, I stated so explicitly in a separate paragraph that follows. I also cross-referenced it from the section on type lambdas.

Reply:
For what it's worth, I was the person calling more things pattern signatures and I have stopped doing that. I am not sure anyone but me made that mistake.